File: clrsCpp/lecture_1/Merge_Sort.cpp
#include <iostream>

int* Merge_operation(int* array, int* left_array, int* right_array,int left_length, int right_length)
{
    int i = 0;
    int j = 0;
    int count = 0;

    while (i < left_length && j < right_length)
    {
        if (*(left_array + i) < *(right_array + j))
        {
            array[count] = *(left_array + i);
            i++;
            count++;
        }
        else
        {
            array[count] = *(right_array + j);
            j++;
            count++;
        }
    }

    if (i < left_length) 
    {
        for (; i < left_length; i++)
        {
            array[count] = *(left_array + i);
            count++;
        }
    }

    if (j < right_length)
    {
        for (; j < right_length; j++)
        {
            array[count] = *(right_array + j);
            count++;
        }
    }
    return array;
}

int* Merge_Sort(int* array, int length)
{
    if (length == 1)
    {
        return array;
    }
    else
    {
        int left_length = length / 2;
        int right_length = length - left_length;

        int* left_array = new int[left_length];
        int* right_array = new int[right_length];
        int count = 0;
        for (int i = 0; i < left_length; i++)
        {
            left_array[i] = *(array + count);
            count++;
        }
        for (int i = 0; i < right_length; i++)
        {
            right_array[i] = *(array + count);
            count++;
        }
        Merge_Sort(left_array, left_length);
        Merge_Sort(right_array, right_length);
        Merge_operation(array,left_array, right_array, left_length, right_length);
        return array;
    }
}

File: clrsPython/Lecture_1/Merge_Sort.py
def Merge_Sort(A):
    length = len(A)
    if length==1:
        return A
    else:
        d = int(len(A)/2)
        A1 = A[0:d]
        A2 = A[d:]
        A1 = Merge_Sort(A1)
        A2 = Merge_Sort(A2)
        return Merge_operation(A1,A2,length)

def Merge_operation(A1,A2,length):
    A = []
    for i in range(length):
        if A1[0] < A2[0]:
            A.append(A1[0])
            if len(A1) == 1:
                A.extend(A2)
                return A
            else:
                A1 = A1[1:]
        else:
            A.append(A2[0])
            if len(A2) == 1:
                A.extend(A1)
                return A
            else:
                A2 = A2[1:]
    return A


A = [9,8,7,6,5,4,3,2,1]
A = Merge_Sort(A)
print(A)